src/lib/gadgets/bitwise.ts

```diff
@@ -0,0 +1,192 @@
+import { Field, toFp } from '../field.js';
+import { Provable } from '../provable.js';
+import { Field as Fp } from '../../provable/field-bigint.js';
+import * as Gates from '../gates.js';
+
+export { xor };
+
+/**
+ * Bitwise XOR gadget on {@link Field} elements. Equivalent to the [bitwise XOR `^` operator in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR).
+ * A XOR gate works by comparing two bits and returning `1` if two bits differ, and `0` if two bits are equal.
+ *
+ * The `length` parameter lets you define how many bits should be compared. The output is not constrained to the length.
+ *
+ * **Note:** Specifying a larger `length` parameter adds additional constraints.
+ *
+ * **Note:** Both {@link Field} elements need to fit into `2^length - 1`, or the operation will fail.
+ * For example, for `length = 2` ( 2² = 4), `.xor` will fail for any element that is larger than `> 3`.
+ *
+ * ```typescript
+ * let a = Field(5);    // ... 000101
+ * let b = Field(3);    // ... 000011
+ *
+ * let c = xor(a, b, 2);    // ... 000110
+ * c.assertEquals(6);
+ * ```
+ */
+function xor(a: Field, b: Field, length: number, lengthXor = 4) {
+  // check that both input lengths are positive
+  assert(
+    length > 0 && lengthXor > 0,
+    `Input lengths need to be positive values.`
+  );
+
+  // check that length does not exceed maximum field size in bits
+  assert(
+    length <= Field.sizeInBits(),
+    `Length ${length} exceeds maximum of ${Field.sizeInBits()} bits.`
+  );
+
+  // sanity check as prover to check that both elements fit into length bits
+  Provable.asProver(() => {
+    assert(
+      a.toBigInt() < 2 ** length,
+      `${a.toBigInt()} does not fit into ${length} bits`
+    );
+
+    assert(
+      b.toBigInt() < 2 ** length,
+      `${b.toBigInt()} does not fit into ${length} bits`
+    );
+  });
+
+  // handle constant case
+  if (a.isConstant() && b.isConstant()) {
+    return new Field(Fp.xor(a.toBigInt(), b.toBigInt()));
+  }
+
+  // calculate expect xor output
+  let outputXor = Provable.witness(
+    Field,
+    () => new Field(Fp.xor(a.toBigInt(), b.toBigInt()))
+  );
+
+  // obtain pad length until the length is a multiple of 4*n for n-bit length lookup table
+  let padLength = length;
+  if (length % (4 * lengthXor) !== 0) {
+    padLength = length + 4 * lengthXor - (length % (4 * lengthXor));
+  }
+
+  // recursively build xor gadget chain
+  buildXor(a, b, outputXor, padLength, lengthXor);
+
+  // return the result of the xor operation
+  return outputXor;
+}
+
+// builds xor chain recursively
+function buildXor(
+  a: Field,
+  b: Field,
+  expectedOutput: Field,
+  padLength: number,
+  lengthXor: number
+) {
+  // if inputs are zero and length is zero, add the zero check
+  if (padLength === 0) {
+    Gates.zeroCheck(a, b, expectedOutput);
+
+    let zero = new Field(0);
+    zero.assertEquals(a);
+    zero.assertEquals(b);
+    zero.assertEquals(expectedOutput);
```

Comment:
I am actually unsure why we need two "zero checks" here, 1) the zero check gate, and 2) asserting that both elements equal zero. Isn't that the same thing?